lib/AST/RequirementMachine/Rule.cpp

```diff
@@ -43,6 +43,33 @@ Optional<Symbol> Rule::isPropertyRule() const {
   return property;
 }
 
+Optional<Requirement> Rule::getPropertyRequirement(Type subject) const {
+  auto property = isPropertyRule();
+  if (!property)
+    return None;
+
+  switch (property->getKind()) {
+  case Symbol::Kind::ConcreteType:
+    return Requirement(RequirementKind::SameType, subject,
+                       property->getConcreteType());
```

Comment:
If the concrete type is generic, getConcreteType() returns a type where type parameters have been replaced with fresh generic parameters. Try something like `T == G<U.T>` vs `T == Int`. You need to use PropertyMap::getTypeFromSubstitutionSchema(), passing in symbol->getConcreteType() and symbol->getSubstitutions().

Reply:
The 'prefix' parameter to getTypeFromSubstitutionSchema() is the prefix of the subject type whose length is the difference in length between the two terms. Suppose you have two conflicting property rules:

- `[P:X].[concrete: G1<[P:Y]>] => [P:X]` from `protocol P { associatedtype X where X == G1<Y>; associatedtype Y }`
- `[Q:X].[P:X].[concrete: G2<[Q:Z]>] => [Q:X].[P:X]` from `protocol Q { associatedtype X : P where X.X == G2<Z>; associatedtype Z }`

The subject type here is the longer of the two right hand sides, `[Q:X].[P:X]`. When converting the concrete type symbol from the first rule `[concrete: G<[P:Y]>]` into a Type, we have to prepend the prefix `[Q:X]` to the substitution term `[P:Y]` to get `[concrete: G<[Q:X].[P:Y]>]`, since the 'adjusted' concrete type requirement is 'X.X == G1<X.Y>` not `X.X == G1<Y>`.